src/api/routes/tokens/tokens.ts

```diff
@@ -23,6 +29,76 @@ export function createTokenRouter(db: DataStore): express.Router {
   const router = express.Router();
   router.use(express.json());
 
+  router.get(
+    '/nft/holdings',
+    asyncHandler(async (req, res, next) => {
+      const principal = req.query.principal;
+      if (typeof principal !== 'string' || !isValidPrincipal(principal)) {
+        res.status(400).json({ error: `Invalid or missing principal` });
+        return;
+      }
+      let assetIdentifiers: string[] | undefined;
+      if (req.query.asset_identifiers !== undefined) {
+        for (const assetIdentifier of [req.query.asset_identifiers].flat()) {
+          if (
+            typeof assetIdentifier !== 'string' ||
+            !isValidPrincipal(assetIdentifier.split('::')[0])
+          ) {
+            res.status(400).json({ error: `Invalid asset identifier ${assetIdentifier}` });
+            return;
+          } else {
+            if (!assetIdentifiers) {
+              assetIdentifiers = [];
+            }
+            assetIdentifiers?.push(assetIdentifier);
+          }
+        }
+      }
+      const limit = parseTokenQueryLimit(req.query.limit ?? 50);
+      const offset = parsePagingQueryInput(req.query.offset ?? 0);
+      const includeUnanchored = isUnanchoredRequest(req, res, next);
+      const includeTxMetadata = booleanValueForParam(req, res, next, 'tx_metadata');
+
+      const { results, total } = await db.getNftHoldings({
+        principal: principal,
+        assetIdentifiers: assetIdentifiers,
+        offset: offset,
+        limit: limit,
+        includeUnanchored: includeUnanchored,
+      });
+      const parsedResults: NonFungibleTokenHolding[] = await Promise.all(
+        results.map(async result => {
```

Comment:
Quite a lot of sql requests used up here. Should the `getNftHoldings` query JOIN against the txs table for these results? Or alternatively, fetch all txs by array in a single additional sql query?

Reply:
Also note this has two separate queries not contained with a a sql transaction, so the first one in `getNftHoldings` can return results for a block that ends up getting forked before the queries afterwards, resulting in either an error or mangled response data.

Reply:
> Quite a lot of sql requests used up here. Should the `getNftHoldings` query JOIN against the txs table for these results? Or alternatively, fetch all txs by array in a single additional sql query?

Interestingly, I found out that performing the JOIN against `txs` in the original query is much more expensive because it creates a lot of duplicate rows which then need to be trimmed by complex query techniques (e.g. once row produced per each `nft_event`<->`tx` possible combination, same problem we have in `/transactions`).

However, I agree with your idea that fetching by list might be the best way to go here. That fetch ends up being a `WHERE tx_id = ANY(x)` which is a much cheaper index lookup. I'll work on this.


> Also note this has two separate queries not contained within a single sql transaction (i.e. `BEGIN; <queries>; COMMIT;`).

Great point. I'll add this too.